Implemented BasePlayer class

4br3mm0rd opened the pull request:
There we go for the BasePlayer class.

I took a lot of the code from the previous MPyg321 class and tried to make an abstract version of anything that could have create a conflict. I also put most of the constants and external classes in external files for clarity.

The main issues that we might find:
- I use a global value for `mpg_outs` and `mpg_codes`. This means that `pexpect` will look for codes that are not supposed to happen in the used version (codes from mpg321 when using mpg123 and codes from mpg123 when using mpg321). I made it for simplicity, and this should not be an issue for now but might be one in the future. 

One of the things that annoys me the most here is the "end_of_song" callback. With mpg321, we have an explicit code for end of song (@P 3) while mpg123 doesn't have any (so I basically implement it in a weird way by using the status `PlayerStatus.STOPPING` value in order to let the class know that it is a user stop and not the end of the song).

Another thing that could be annoying here is the fact that the class is ready to call for `on_user_mute`/`on_user_unmute` while it doesn't make sense for mpg321.

An easy fix would be to implement the process_output class in the subclasses instead of the main one... But this is such a central class that I find it a bit sad. What do you guys think?

- Another thing that is annoying is the "-R mpyg" argument. When using mpg123, it seams that the "mpyg" part of the argument makes it impossible to add other arguments after it. This is the reason why I inverted the `--audiodevice` and the `-R` args so that it works (I haven't tested it though, so it is worth testing).

Same problem here, if we want to implement it in the cleanest way, we should externalize this very central part of the code.

Waiting for your comments :)

4br3mm0rd committed:
Implemented BasePlayer class

4br3mm0rd commented:
@hakmad @stefets

stefets commented:
@4br3mm0rd good job!

Let me share my comments!

> One of the things that annoys me the most here is the "end_of_song" callback.

I agree, and this event make big sense... we have few choices. What do you think about a feature request ? https://sourceforge.net/p/mpg123/feature-requests/ 
We would have the opportunity to present the project and at the same time ask for an explicit code when a song ends.

> Another thing that could be annoying here is the fact that the class is ready to call for on_user_mute/on_user_unmute

Should the on_user_mute / on_user_unmute methods be implemented in the class that will implement mpg123? Or throw some kind of METHOD NOT ALLOWED exception if mpg321 ?

> Another thing that is annoying is the "-R mpyg" argument.

Child classes could pass the arguments to the base class, which takes care of instantiating the right process with the appropriate collection of arguments.

Meaning the constructor would look something like `BasePlayer.__init__(self, **kwargs)` and pass the kwargs to the pexpect.spawn

Thanks!

4br3mm0rd commented:
Thanks @stefets 

> I agree, and this event make big sense... we have few choices. What do you think about a feature request ? https://sourceforge.net/p/mpg123/feature-requests/
> We would have the opportunity to present the project and at the same time ask for an explicit code when a song ends.

Great idea! I'll submit it :) 

> Should the on_user_mute / on_user_unmute methods be implemented in the class that will implement mpg123? Or throw some kind of METHOD NOT ALLOWED exception if mpg321 ?

That could be dangerous. Since this is a callback based on the output of mpg123, this could throw an exception at unexpected times.
I realize now that this is also a problem with my implementation so we should think of another way to implement it!

Should we put the `output_processor` inside the sublcasses? Or can you think of another way?

> Child classes could pass the arguments to the base class, which takes care of instantiating the right process with the appropriate collection of arguments.
> 
> Meaning the constructor would look something like BasePlayer.__init__(self, **kwargs) and pass the kwargs to the pexpect.spawn

I like that!

4br3mm0rd commented:
[Feature request is here](https://sourceforge.net/p/mpg123/feature-requests/65/) :)

stefets commented:
Hello @4br3mm0rd,
 
> Feature request is here :) 

Cooool!

For the feature request i have check the source code, it's in control_generic.c and it's in the else statement, i think.
```
/* When the track ended, user may want to keep it open (to seek back),
   so there is a decision between stopping and pausing at the end. */
if(param.keep_open)
{
	mode = MODE_PAUSED;
	/* Hm, buffer should be stopped already, shouldn't it? */
	generic_sendmsg("P 1");
}
else
{
	mode = MODE_STOPPED;
	close_track();
	generic_sendmsg("P 0");
}
```


> Should we put the output_processor inside the sublcasses? Or can you think of another way?

I like having the output_processor in the bass class and I feel confortable having the mute/unmute in the BasePlayer because of it's 'hybrid' state. 

What do you think about having a common mpg_outs but having a specific list for each player and extend this list in the child class ? This will be more explicit and we can document in the BasePlayer code that the mute/onmute is 123 specific and have no effects when using 321.

Example:
```
mpg_outs_123 = [
    {
        "mpg_code": "@mute",
        "action": "user_mute",
        "description": "Player has been muted by the user."
    },
    {
        "mpg_code": "@unmute",
        "action": "user_unmute",
        "description": "Player has been unmuted by the user."
    },
]

mpg_outs_321 = [
    {
        "mpg_code": "@P 3",
        "action": "end_of_song",
        "description": "Player has reached the end of the song."
    },
]

```

So the child class mpg123(BasePlayer) would extend the list.

`mpg_outs.extend(mpg_outs_123)`

Thanks!

4br3mm0rd commented:
@stefets 
Thanks for the input, I'll work on this :) 
Can you check the implementation described here: https://sourceforge.net/p/mpg123/feature-requests/65/#ce0a ?

I'm not sure how to test it

stefets commented:
@4br3mm0rd
It works, we will have to deal with a P 0 that happen immediatly after the P 3 message

You can see changes in the svn mirror here : 
- https://github.com/madebr/mpg123/commit/7c2971250233ea945a06f8c63e1aefe7cc443816

I have downloaded the snapshot and build/install and here is the output results:

$ mpg123 --version
mpg123 20220117022201

$ mpg123 --remote
@R MPG123 (ThOr) v10
silence
@silence
load 0-theme.mp3
@I {
@I }
@P 2
@S 2.0 3 22050 Single-Channel 0 156 1 0 0 0 48 0 0
@P 3
@P 0
quit
$

At the same time, a new remote command will be available : PROGRESS
- @H SILENCE: be silent during playback (no progress info, opposite of PROGRESS)
- @H PROGRESS: turn on progress display (opposite of SILENCE)

FYI about how I have made this test 
* Download the snapshot https://www.mpg123.de/snapshot or clone the mirror https://github.com/madebr/mpg123 (master)
* Follow how I build mpg123 here https://github.com/stefets/live-config/wiki/How-I-build-MPG123

We have it !

4br3mm0rd committed:
Implemented extended outputs in BaseClass

4br3mm0rd committed:
Implemented MPyg32Player class + bugfixes in BasePlayer + adapted examples

4br3mm0rd committed:
Removed on_user_mute/unmute functions from process_output - should be implemented in the MPyg123Player child class

4br3mm0rd committed:
Added missing MPyg321Player class

4br3mm0rd committed:
Remove useless import

4br3mm0rd commented:
Hi @stefets !

So here is what I did (sorry for the delay).
- I implemented the process_output_ext for the output that should be handled in the child class. 
- I changed the `check_player` function because it had some issues on macOS when using pexpect ([see here](https://pexpect.readthedocs.io/en/stable/commonissues.html#truncated-output-just-before-child-exits)) and actually we don't need pexpect here so... all good :) 
- I did NOT implement the `@P 3` opcode... The reason why I made this choice is because I have been downloading the latest versions of mpg123 on MacOS Homebrew and on Ubuntu's apt-get and none of them had a version that supported it, so I think supporting it right now will only introduce new bugs.


Let me know what you think :)

stefets commented:
Hi @4br3mm0rd ! 

Glad to hear news from you. The BasePlayer make sense, good job!

About the P3 opcode and the mpg123 implementation I have a suggestion to support it.

There is a reason why the mpg123 version provided by Homebrew or Aptitude is not the version supporting the new P3 code. I think it's because the application (and many others) is not in the 'pipeline' of the OS updates.

Some users will install mpg123 with OS package manager and other users, like me, will compile the source code to get the latest mpg123 binaries, and the P3 code as well.

Because of that, there is a breaking change and I propose to check the mpg123 version in the BasePlayer with a regex.

If the mpg123 version is greater or equal than the version that implement the P3 opcode, the BasePlayer (or the child class) support it, else we do nothing and there is no breaking change.

Does it make sense ?

stefets commented:
Hi again :)

If you check this commit https://github.com/madebr/mpg123/commit/7c2971250233ea945a06f8c63e1aefe7cc443816

The P 3 code will be in the 1.30.0 release.

I agree with you to don't implement it at this moment. I can do some tests on my side by compiling the snapshot that contain the modification.

Thanks!

4br3mm0rd commented:
Hey @stefets !

Thanks for the answer.

> If the mpg123 version is greater or equal than the version that implement the P3 opcode, the BasePlayer (or the child class) support it, else we do nothing and there is no breaking change.

I think we should actually remove opcode `@P 3` from our system (at least for now), simply because the implementation we have right now is working and is cross-versions.

Also, handling versions will make the code more complex, and the fact that `@P 0` will still be sent in mpg123 will still make the code a bit different from mpg123 to mpg321.

If we remove opcode `@P 3` from the system, we will have a unified code for mpg321 and mpg123 which is basically what we are trying to do. So, obviously, this code looks a bit more messy than it would have been with `@P 3`  (especially because of the `STOPPING` status), but I think it is still the cleaner version.

What do you think?

stefets commented:
Hi @4br3mm0rd!

Indeed, in this meaning it's all abstract and the mpg123 legacy wrapper is very simple

As a user, in my plugin the change is simple as hell

`class Mp3Player(MPyg321Player):` to `class Mp3Player(BasePlayer):` that's satisfying for me :+1: 

This is a great improvement :1st_place_medal: 

Should we cancel our request to add the P3 code in the mpg123 application ? 

Thanks !

4br3mm0rd commented:
> As a user, in my plugin the change is simple as hell
> 
> `class Mp3Player(MPyg321Player):` to `class Mp3Player(BasePlayer):` that's satisfying for me 👍

As we said in #22 , we need to implement `MPyg321Player` and `MPyg123Player` classes that will inherit from the `BasePlayer` class.

> Should we cancel our request to add the P3 code in the mpg123 application ?
> 
> Thanks !

I guess there is no need for that, since it's already in the code...

I'll merge this PR since you seem to be ok with it (merge to branch v2.0 - we need the subclasses before we merge to master and release).
